lib/teleterm/gateway/gateway_test.go

```diff
@@ -45,3 +50,42 @@ func TestCLICommandUsesCLICommandProvider(t *testing.T) {
 
 	require.Equal(t, "foo/bar", command)
 }
+
+func TestGatewayStart(t *testing.T) {
+	hs := httptest.NewTLSServer(http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {}))
+	t.Cleanup(func() {
+		hs.Close()
+	})
+
+	gateway, err := New(
+		Config{
+			TargetName:   "foo",
+			TargetURI:    uri.NewClusterURI("bar").AppendDB("foo").String(),
+			TargetUser:   "alice",
+			Protocol:     defaults.ProtocolPostgres,
+			CertPath:     "../../../fixtures/certs/proxy1.pem",
+			KeyPath:      "../../../fixtures/certs/proxy1-key.pem",
+			Insecure:     true,
+			WebProxyAddr: hs.Listener.Addr().String(),
+		},
+		mockCLICommandProvider{},
+	)
+	require.NoError(t, err)
+
+	serveErr := make(chan error)
+
+	go func() {
+		err := gateway.Serve()
+		serveErr <- err
+	}()
+
+	// Dial to make sure gateway is open.
+	gatewayAddress := net.JoinHostPort(gateway.LocalAddress, gateway.LocalPort)
+	conn, err := net.DialTimeout("tcp", gatewayAddress, time.Second*1)
+	require.NoError(t, err)
```

Comment:
@espadolini One thing that worries me is that this check fails only if I don't start the listener on that address. It doesn't fail if I never actually start the local proxy or if I start it after 1.5 second.

It turns out that `localProxy.Start()`, [unlike gRPC's `Server.Serve()`](https://github.com/grpc/grpc-go/blob/v1.46.0/server.go#L737-L746), doesn't error if it's called after `Close()`.

So the test not failing if `Serve()` is called after `Close()` is okay I guess. But why doesn't the test fail if never start the local proxy?

Reply:
I tried to adapt this to [`teleterm_test.go`](https://github.com/gravitational/teleport/blob/00d7070fef052cf08f3fe4ff1ca78a8963afe139/lib/teleterm/teleterm_test.go) but it seems I misunderstood how `Dial` works. It doesn't actually wait for the other side to accept the connection since, as you said, connections are buffered as soon as the listener is started.

So I cannot really use `Dial` there to make sure the server is accepting connections.

<details>
<summary>Diff of teleterm_test.go</summary>

```diff
diff --git a/lib/teleterm/teleterm.go b/lib/teleterm/teleterm.go
index b5356fdc9..123a5be95 100644
--- a/lib/teleterm/teleterm.go
+++ b/lib/teleterm/teleterm.go
@@ -16,8 +16,10 @@ package teleterm
 
 import (
 	"context"
+	"fmt"
 	"os"
 	"os/signal"
+	"time"
 
 	"github.com/gravitational/teleport/lib/teleterm/apiserver"
 	"github.com/gravitational/teleport/lib/teleterm/clusters"
@@ -60,7 +62,10 @@ func Start(ctx context.Context, cfg Config) error {
 
 	serverAPIWait := make(chan error)
 	go func() {
+		time.Sleep(time.Second * 2)
+		fmt.Println("apiServer.Serve()")
 		err := apiServer.Serve()
+		fmt.Println("apiServer.Serve() returned")
 		serverAPIWait <- err
 	}()
 
diff --git a/lib/teleterm/teleterm_test.go b/lib/teleterm/teleterm_test.go
index 51be1b619..649bac66f 100644
--- a/lib/teleterm/teleterm_test.go
+++ b/lib/teleterm/teleterm_test.go
@@ -16,7 +16,11 @@ package teleterm_test
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"net"
+	"os"
+	"path/filepath"
 	"testing"
 	"time"
 
@@ -27,27 +31,38 @@ import (
 
 func TestStart(t *testing.T) {
 	homeDir := t.TempDir()
+	sockPath := filepath.Join(homeDir, "teleterm.sock")
+	addr := fmt.Sprintf("unix://%v", sockPath)
 	cfg := teleterm.Config{
-		Addr:    fmt.Sprintf("unix://%v/teleterm.sock", homeDir),
+		Addr:    addr,
 		HomeDir: fmt.Sprintf("%v/", homeDir),
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
-	wait := make(chan error)
+	serveErr := make(chan error)
 	go func() {
 		err := teleterm.Start(ctx, cfg)
-		wait <- err
+		serveErr <- err
 	}()
 
-	defer func() {
-		// Make sure Start() is called.
-		time.Sleep(time.Millisecond * 500)
+	require.Eventually(t, func() bool {
+		_, err := os.Stat(sockPath)
 
-		// Stop the server.
-		cancel()
-		require.NoError(t, <-wait)
-	}()
+		return !errors.Is(err, os.ErrNotExist)
+	}, time.Millisecond*500, time.Millisecond*50)
+
+	// Dial to make sure Start was called.
+	fmt.Println("net.DialTimeout()")
+	conn, err := net.DialTimeout("unix", sockPath, time.Second*1)
+	fmt.Println("net.DialTimeout() returned")
+	require.NoError(t, err)
 
+	// Stop the server.
+	fmt.Println("cancel()")
+	cancel()
+	conn.Close()
+	fmt.Println("waiting for serveErr")
+	require.NoError(t, <-serveErr)
 }
```
</details>

Reply:
Okay, I think `conn.Read()` does the trick. I just pushed 3e34e0208 which fails if the connection isn't accepted within a second.

I was also able to improve teleterm tests in a similar way, I'll push those changes in a separate PR.